src/Microsoft.HttpRepl/OpenApi/OpenApiDotNetApiDefinitionReader.cs

```diff
@@ -0,0 +1,108 @@
+
+        public ApiDefinition ReadDefinition(string document, Uri sourceUri)
+        {
+            OpenApiStringReader reader = new();
+
+            OpenApiDocument openApiDocument = reader.Read(document, out _);
+
+            ApiDefinition apiDefinition = new ApiDefinition();
+            List<EndpointMetadata> metadata = new List<EndpointMetadata>();
+
+            foreach (OpenApiServer server in openApiDocument.Servers)
+            {
+                string url = server.Url?.EnsureTrailingSlash();
+                string description = server.Description;
+
+                if (url is null)
+                {
+                    continue;
+                }
+
+                if (Uri.IsWellFormedUriString(url, UriKind.Absolute) && Uri.TryCreate(url, UriKind.Absolute, out Uri absoluteServerUri))
+                {
+                    apiDefinition.BaseAddresses.Add(new ApiDefinition.Server() { Url = absoluteServerUri, Description = description });
+                }
+                else if (Uri.TryCreate(sourceUri, url, out Uri relativeServerUri))
+                {
+                    apiDefinition.BaseAddresses.Add(new ApiDefinition.Server() { Url = relativeServerUri, Description = description });
+                }
+            }
+
+            if (openApiDocument.Paths is not null)
+            {
+                foreach (KeyValuePair<string, OpenApiPathItem> path in openApiDocument.Paths)
+                {
+                    string relativeUrl = path.Key;
+
+                    EndpointMetadata endpointMetadata = new EndpointMetadata(relativeUrl);
+
+                    foreach (KeyValuePair<OperationType, OpenApiOperation> operation in path.Value.Operations)
+                    {
+                        RequestMetadata requestMetadata = new RequestMetadata(operation.Key);
+
+                        foreach (OpenApiParameter parameter in operation.Value.Parameters)
+                        {
+                            requestMetadata.Parameters.Add(parameter);
+                        }
+
+                        if (operation.Value.RequestBody?.Content is not null)
+                        {
+                            foreach (KeyValuePair<string, OpenApiMediaType> content in operation.Value.RequestBody.Content)
+                            {
+                                string contentType = content.Key;
+                                bool isRequired = operation.Value.RequestBody.Required;
+                                OpenApiSchema schema = content.Value?.Schema;
+
+                                requestMetadata.Content.Add(new RequestContentMetadata(contentType, isRequired, schema));
+                            }
+                        }
+
+                        endpointMetadata.AvailableRequests.Add(requestMetadata);
+                    }
+
+                    metadata.Add(endpointMetadata);
+                }
+            }
+
+            DirectoryStructure d = new DirectoryStructure(null);
```

Comment:
This could also be pulled out into a separate utility method, e..g `apiDefinition.DirectoryStructure = BuildDirectoryStructure(metadata);`